### ISSUE TITLE
`cat | cat <no_such_file >/no_permission` breaks Fish

### ISSUE THREAD

**Issue by jos128** [fish-shell/fish-shell#7411]
Likely related to #7038.

I stumbled into another flavour of the issue with IO redirection. Fish session in *Gnome Terminal*:
```console
➜ ~ sh -c 'env HOME=/tmp/fish_tmp fish'
```
```console
$ cat | cat < no_such_file
<W> fish: An error occurred while redirecting file 'no_such_file'
open: No such file or directory
$ ⏎ 
```
**Crash at error on source redirection**.

CPU fan indicates a problem. `htop` shows one `fish` processes at 100% CPU load.
```console
➜ ~ jobs -l
Job	Group	CPU	State	Command
1	502	0%	stopped	sh -c 'env HOME=/tmp/fish_tmp fish'
➜ ~ ps -o pid,pcpu,stat,comm
  PID %CPU STAT COMMAND
  354  0.1 Ssl  fish
  502  0.0 T    sh
  503 87.3 R    fish
  526  0.0 Z    cat <defunct>
```
```console
➜ ~ kill 503 
➜ ~ ps -o pid,pcpu,stat,comm
  PID %CPU STAT COMMAND
  354  0.2 Ssl  fish
  502  0.0 T    sh
  503 22.7 Z    fish <defunct>
```

---
### Notes
* With file descriptor duplication error on source redirection seems to be fine, though:
```console
$ cat | cat <&1 no_such_file
cat: no_such_file: No such file or directory
I can write one line like this...
$
```
* Of course, this issue happens without the subshelling, too, but then the terminal window dies and takes everything with it.
* `bash` handles the problem fine.
* I realized the output redirection is exactly #7038 and there is no new `fish` version still, so I only keep  the input problem open now, as possible thing to test the fix against. Can't edit the title. Sorry for the spam. 

### Info
```
fish, version 3.1.2 (PPA)
xterm-256color
5.4.0-51-generic x86_64 GNU/Linux
```

**Comment by krobelus** [comment]
On current master, `cat | cat < no_such_file` prints the same error (`An error occurred while redirecting file 'no_such_file'`) but then reads from stdin. This behavior seems correct, especially given that it's a borderline invalid construct because the second `cat` has both piped and redirected input. On 3.1.2 it exposes the, now fixed 100% CPU problem. 

> $ cat | cat <&1 no_such_file
cat: no_such_file: No such file or directory
I can write one line like this...

This is interesting. I guess the job could stop right away, before reading a line.

**Comment by jos128** [comment]
> This behavior seems correct, especially given that it's a borderline invalid construct because the second cat has both piped and redirected input.

Yes, but I stumbled into this organically, just reduced it into something basic with `cat`. Happens, if you are trying `bash` things, e.g. `cat A | diff - <(cat B)` (this breaks fish, if B contains no valid filenames). 

To be fair `<&1 ...` does not do, what I thought it does. It just worked for me with `cat`, because `cat file1 file2` is valid. Sorry, for putting up with my "helping" XD

> This is interesting.

`bash` does the same and apparently it has nothing to do with the `<&1` loop, as `cat | true` does the same.

While playing with this I also noticed `cat file > fifo &` does not background the job. Is that supposed to happen?

**Comment by mqudsi** [comment]
As @krobelus mentioned this does expose a bug in the old version of fish, but aside from that (fixed) issue, I don't see anything wrong or anything that can be done differently.

Some things to note:

* `foo <&1 doesnt_exist` doesn't try to read from `doesnt_exist`, that's a redirect from fd 1 (stdin) concatenated with an invalid input.
* krobelus was being charitable, `cat | cat < foo` isn't borderline invalid, it's just invalid. fish mimics bash here and just drops the `< foo` because there's already an input, but tbh the most correct thing to do would be throw a parser error and fail to even execute.
* `cat | foo` where `foo` exits immediately will always wait for a single line to be entered because `cat` is (at least partially) line buffered and has control of the shell; it is reading from a valid input fd (stdin, which is echoing back what you type due to the tty flags) and hasn't tried to write to the now-broken output pipe connecting it to `foo` (which already exited). It's only when it tries to write to the output fd and finds it closed that it exits.
* `cat` expects the output can disconnect so it doesn't error/warn when it finds the output fd closed, but when you start cat with the output fd closed it will refuse to execute: `cat <&-`
* After fish drops the invalid second input to the second cat already reading from the first cat, you are in effect only executing  `cat | cat does_not_exist`, which is (as far as the first cat is concerned) equal to `cat | true`, i.e. the first cat should read from stdin and write to the output fd which starts off connected to the next command in the chain (here, `true) so it passes the initial sanity check, but as soon as it tries to write the first line-buffered input to the output, it finds the output fd closed and terminates.

You can test how much of this you understand by guessing the behavior of this very simple case (it's perfectly valid and not a trick or very hard question):

```
cat | sleep 10
```

---

All that said, your fifo find is interesting. The job doesn't background because fish hangs on the redirect (before the `&` comes into the picture). A fifo blocks when the pipe is full, until the reader empties it sufficiently (under the high water mark). Fish doesn't typically sit in the middle between producer and consumer so it shouldn't be hanging when it's just opening the fd to dup it - it should be `cat` that's hanging, not fish. On my PC, interrupting that with a ^C causes a warning about an interrupted syscall and if the command was being backgrounded, then fish dies with an assertion failure:

```
fish: ../src/exec.cpp:1050: bool exec_job(parser_t &, const shared_ptr<job_t> &, const io_chain_t &): Assertion `pgid.has_value() && "Background jobs should always have a pgroup"' failed.
fish: Job 1, 'fish' terminated by signal SIGABRT (Abort)
```

which seems like it is assuming the job was created when redirection failed and losing sanity thereafter (in a separate issue or bug from the hang).

**Comment by jos128** [comment]
Yeah, I am totally embarrassed by the fact, that I re-reported my own first fish issue. I was somehow convinced, that was gone by now. Again, sorry for the noise.
 
Thank you, very much for the detailed insights! Really appreciated. If you don't mind: How would I go about finding out these things for myself? Does fish have a "meta mode" to observe what's going on? (If not, that would be really cool!) Which programs would allow to do so otherwise? And is there a way to mix/multiplex `stdin` redirections in fish? It works for `stdout` after all.  

I agree it would be helpful for fish to throw a parsing error, if it's invalid. For a beginner it's not obvious why `cat A | cat B` prints `B`, but `cat A | cat < B` prints `A` (*I* get it now). If I understood you correctly, the first case can't really be detected by the shell (since it's valid and the first `cat`'s issue), but at least the second one should be possible, right? 

I stumbled into the fifo bug playing with redirections, to find out what the differences are between the "various" redirections (as you may have guessed I am not a RTFM type of person).  In `bash` I would do this:

```console
> echo hello > fifo &
[1] 1234
> cat fifo
hello
[1]+  Done                    echo hello > fifo
```
Glad, I could at least break this one XD

In `fish 3.1.2` the issue doesn't seem to be related to the `&`, as `echo hello > fifo` seems to hang in `fish` not `echo` as well. `CTRL-C` prints `^C<W> fish: An error occurred while redirecting file 'fifo'
open: Interrupted system call`, same as with the `... &` case.

**Comment by mqudsi** [comment]
No worries, it's always fun to dig into this stuff every once in a while, and you never know who will google some relevant terms at some point in the future and learn from everything we are posting.

> For a beginner it's not obvious why `cat A | cat B` prints `B`
> the first case can't really be detected by the shell (since it's valid and the first cat's issue)

There isn't an issue at all actually, you just have to take a step back to see what the shell's role in all this is. The shell really does nothing more than launch process groups and marshal the input/output between the processes. Each process has one input (stdin) and two outputs (stdout and stderr). A shell pipeline is complete when there is no dangling output left. The inputs/outputs are what connects pipeline members (processes) to one another. If you go back to the invalid construct: `foo1 | foo2 < foo3`, dropping `< foo3` is actually the only move available to the shell because if `foo2` were connected to input from `foo3` there would be nothing connecting `foo1` to `foo2`. It would be akin to running `foo1 &; foo2 < foo3 &;` then waiting on both to exit (so really they're just separate jobs with nothing shared between them), whereas so long as there is an open pipe between the processes their lifetimes are (indirectly) tied to one another. It's why `cat huge_file | head -n 1` can exit so fast even if `cat huge_file` would normally take hours to complete - once the output pipe for the first process has closed, it knows to exit.

Back to your example, `cat A | cat B` has the first cat emitting output that is marshalled (by fish) to the stdin of the second cat, but the second cat completely disregards its stdin because that's just how cat works - the stdin is only read if no source is provided (or if the magic source `-` is provided). The output of the first cat doesn't just "go nowhere" because the output is a pipe, meaning once the buffer is full the first `cat` will actually hang waiting for the buffer to drain to be able to write more, but since the second `cat` never reads from it, it just ends up exiting when the fd is closed, which happens when the second cat exits.

The fifo bug is tricky. The open syscall on the fifo purposely hangs because the actual kernel pipe is only around so long as there is a reader or a writer connected, meaning if it didn't block and the process wrote less than the pipe buffer, the process would exit immediately and the output it generated would be lost, even if the very next command was a process opening the fifo to read from it. I'm opening a separate issue for this.

**Comment by jos128** [comment]
Thank you again, for taking the time to explain the internals!

**Comment by mqudsi** [comment]
@jos128 thank you very much for finding the fifo hang problem. The issues raised in this thread are followed up on in #7422 and #7423; I'm going to close out this issue as it doesn't seem that there's anything else that's actionable here. If you have any followup questions or want to continue the discussion, please feel free to continue commenting here and we'll be more than happy to respond - this just makes it clearer that we've dealt with the GitHub issue so we don't get bogged down with too many open reports.
